### DISCUSSION TITLE
Get distinct attribute values in timeframe?

### DISCUSSION THREAD

**Discussion by ovidiubuligan** [grafana/tempo#2236] [Q&A] [answered]
I am interested in a odd query that returns distinct `resource.service.instance.id`  for a 1h timeframe.  It is a "_what services were active ?_" when some problem happened . 
Must Traceql always returns a list of spans by design  ? 
I am not sure it it is in the scope of traceql  , but is it on the roadmap with aggregates ?

**Comment by joe-elliott** [comment, marked as answer]
> Must Traceql always returns a list of spans by design ?

As it's designed now, yes. TraceQL always returns a set of spans. There is a long term goal that it can also return metrics but that is not on the immediate roadmap. We are, however, currently just beginning work on a more dynamic way to generate metrics from spans (more dynamic then the canned metrics from metrics-generator). I will @mdisibio b/c he is starting this work and may be able to share more.

> I am not sure it it is in the scope of traceql , but is it on the roadmap with aggregates ?

TraceQL aggregates will still be built around the concept of selecting traces. For instance if you wanted to find a trace that had more than 1 error in any namespace you could write something like:

```
by(resource.namespace) | { status = error } | count() > 1
```

**Comment by ovidiubuligan** [comment]
@joe-elliott , maybe related functionality , how does autocomplete in grafana TraceQL UI  list the distinct values for an attribute ?

**Reply by joe-elliott** [reply]
This is a good point. You could hit this endpoint to get recent, unique `resource.service.instance.id`:

```
$ curl http://tempo:3200/api/v2/search/tag/resource.service.instance.id/values
```

This will give you all unique instance ids still in the ingesters (wal and all local complete blocks). However, it does not accept time range parameters and will not read from blocks in object storage.
